File: src/main/java/com/gmail/borlandlp/minigamesdtools/arena/team/lobby/spectator/ExampleSpectatorLobby.java
package com.gmail.borlandlp.minigamesdtools.arena.team.lobby.spectator;

import com.gmail.borlandlp.minigamesdtools.Debug;
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools;
import com.gmail.borlandlp.minigamesdtools.arena.ArenaEventListener;
import com.gmail.borlandlp.minigamesdtools.arena.team.lobby.ArenaLobby;
import com.gmail.borlandlp.minigamesdtools.arena.team.lobby.PlayerLocker;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

public class ExampleSpectatorLobby extends ArenaLobby implements SpectatorLobby, PlayerLocker {
    private Set<Player> players = new HashSet<>();
    private boolean enabled_watching;
    private ArenaEventListener listener;

    public boolean isEnabled_watching() {
        return enabled_watching;
    }

    public void setEnabled_watching(boolean enabled_watching) {
        this.enabled_watching = enabled_watching;
    }

    @Override
    public void onInit() {
        this.listener = new ExampleSpectatorLobbyListener(this);
        Bukkit.getServer().getPluginManager().registerEvents((Listener) this.listener, MinigamesDTools.Companion.getInstance());
    }

    @Override
    public void beforeGameStarting() {

    }

    @Override
    public void gameEnded() {
        HandlerList.unregisterAll((Listener) this.listener);
        for (Player player : new ArrayList<>(this.players)) {
            this.removePlayer(player);
        }
    }

    @Override
    public void update() {

    }

    @Override
    public void beforeRoundStarting() {

    }

    @Override
    public void onRoundEnd() {

    }

    @Override
    public void forceReleasePlayer(Player p) {
        this.removePlayer(p);
    }

    @Override
    public void addPlayer(Player player) {
        Debug.print(Debug.LEVEL.NOTICE, "Add player[name:" + player.getName() + "] to Spectator lobby, arena:" + this.getTeamProvider().getArena().getName());
        player.teleport(this.getSpawnPoint());
        this.players.add(player);

        if(MinigamesDTools.Companion.getInstance().getHotbarAPI().isBindedPlayer(player)) {
            MinigamesDTools.Companion.getInstance().getHotbarAPI().unbindHotbar(player);
        }
        if(this.isHotbarEnabled()) {
            MinigamesDTools.Companion.getInstance().getHotbarAPI().bindHotbar(this.getHotbarFor(player), player);
        }
    }

    @Override
    public void removePlayer(Player player) {
        this.players.remove(player);
        /*if(player.getSpectatorTarget() != null) {
            player.setSpectatorTarget(null);
        }
        player.setGameMode(GameMode.SURVIVAL);*/

        if(this.isHotbarEnabled()) {
            MinigamesDTools.Companion.getInstance().getHotbarAPI().unbindHotbar(player);
        }
        if(this.getTeamProvider().getArena().getHotbarController().isEnabled()) {
            try {
                MinigamesDTools.Companion.getInstance().getHotbarAPI().bindHotbar(this.getTeamProvider().getArena().getHotbarController().buildDefaultHotbarFor(player), player);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public Set<Player> getPlayers() {
        return this.players;
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/arena/team/lobby/spectator/ExampleSpectatorLobby.kt
package com.gmail.borlandlp.minigamesdtools.arena.team.lobby.spectator

import com.gmail.borlandlp.minigamesdtools.Debug
import com.gmail.borlandlp.minigamesdtools.Debug.print
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools.Companion.instance
import com.gmail.borlandlp.minigamesdtools.arena.ArenaEventListener
import com.gmail.borlandlp.minigamesdtools.arena.team.lobby.ArenaLobby
import com.gmail.borlandlp.minigamesdtools.arena.team.lobby.PlayerLocker
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener
import java.util.*

class ExampleSpectatorLobby : ArenaLobby(), SpectatorLobby, PlayerLocker {
    override val players: MutableList<Player> = mutableListOf()
    var isEnabled_watching = false
    private var listener: ArenaEventListener? = null

    override fun onInit() {
        listener = ExampleSpectatorLobbyListener(this)
        Bukkit.getServer().pluginManager
            .registerEvents(listener as Listener?, instance)
    }

    override fun beforeGameStarting() {}
    override fun gameEnded() {
        HandlerList.unregisterAll(listener as Listener?)
        for (player in ArrayList(players)) {
            removePlayer(player)
        }
    }

    override fun update() {}
    override fun beforeRoundStarting() {}
    override fun onRoundEnd() {}
    override fun forceReleasePlayer(p: Player) {
        removePlayer(p)
    }

    override fun addPlayer(player: Player) {
        print(
            Debug.LEVEL.NOTICE,
            "Add player[name:" + player.name + "] to Spectator lobby, arena:" + teamProvider!!.arena.name
        )
        player.teleport(spawnPoint)
        players.add(player)
        if (instance!!.hotbarAPI!!.isBindedPlayer(player)) {
            instance!!.hotbarAPI!!.unbindHotbar(player)
        }
        if (this.isHotbarEnabled) {
            instance!!.hotbarAPI!!.bindHotbar(getHotbarFor(player)!!, player)
        }
    }

    override fun removePlayer(player: Player) {
        players.remove(player)
        /*if(player.getSpectatorTarget() != null) {
            player.setSpectatorTarget(null);
        }
        player.setGameMode(GameMode.SURVIVAL);*/if (this.isHotbarEnabled) {
            instance!!.hotbarAPI!!.unbindHotbar(player)
        }
        if (teamProvider!!.arena.hotbarController!!.isEnabled) {
            try {
                instance!!.hotbarAPI!!.bindHotbar(
                    teamProvider!!.arena.hotbarController!!.buildDefaultHotbarFor(
                        player
                    ), player
                )
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
}